### ISSUE TITLE
trajectory controller succeeds too early  (ros ticket #3541)

### ISSUE THREAD

**Issue by ahendrix** [PR2/pr2_controllers#262]
The trajectory controller sometimes succeeds when it is going through the goal tolerance region. However, after declaring success it continues to move and oscillates for a while. The controller should only declare success for stationary goals if its joint velocities are also below a threshold.

trac data:
- Owner: **sglaser**
- Reporter: **sachinc**
- Reported at: **Wed Jan 13 12:32:42 2010**
- URL: https://code.ros.org/trac/wg-ros-pkg/ticket/3541

**Comment by ahendrix** [comment]
[gunter] I would strongly suggest tying in (a) the trajectory should be done.  (b) possibly waiting a short "settling time", though a smooth trajectory shouldn't need that.  (c) at that point, report success if pos and vel are within limits or error if not.

Alternatively, I think we talked about having the controller report "good/bad" if tracking is within expected tolerances.  Which means if all is good, then you can directly report success simply when the trajectory is done??

Well, this begs the questions of WHY is the arm still moving and oscillating at the end of a trajectory??

-gunter

**Comment by ahendrix** [comment]
[sachinc] There's two issues here

(a) The arm moving and oscillating at the end of a trajectory - the gains are badly tuned

(b) The controller saying its done when its really not - we do have a short settling time but without checking on velocity errors, the settling time does not solve the problem.

**Comment by ahendrix** [comment]
[sglaser] Trajectory action no longer succeeds when the position is reached but the velocity error is still high.

r29353

**Comment by ahendrix** [comment]
[sachinc] This does not seem to be working. Matei (who is cc'ed) on this has been using this with move_arm and the controller still comes back with success and then the arm oscillates a little more.

**Comment by ahendrix** [comment]
[sglaser] Matei, were you using "latest" or "trunk"?  (The fix hasn't made it into "latest" yet).

**Comment by ahendrix** [comment]
[matei] I just switched to trunk. I still see the same (quite large) oscillations at the end...

**Comment by ahendrix** [comment]
[sglaser] Does the action succeed?  If so, when does it succeed?  Before or after the oscillations stop?

**Comment by ahendrix** [comment]
[sglaser] Removing milestone.  We will fix this in a patch release.

**Comment by ahendrix** [comment]
[matei] As far as I can tell, when the action succeeds, it does so before the oscillations stop.

Stu & I looked at it some more and it seems the largest oscillations are in the shoulder pan joint. I have attached a plot for accelerations (desired), velocities (desired and actual) and positions (desired and actual) for that joint.

**Comment by ahendrix** [comment]
[matei] We've isolated and recorded a couple of arm trajectories, to make debugging of the oscillations problem easier. There's also a small executable for playing back those trajectories.

Everything needed to reproduce the behavior is inside the package called "test_trajectory":
- rosmake the package
- start the executable bin/test_node, which just listens for and relays joint trajectories to the Joint Trajectory Action
- in bags/ there are two bag files. rosplay either of them and the arm should move

Notes:
- each bag file has two desired trajectories (move the arm in front of the robot, and then out to the side).
- each bag file assumes the arm _starts_ out on the side, and then also leaves the arm there when finished. If the arm is not out to the side to begin with, play any bag to get it there, then play the bag again for the intended behavior

Both bag files contain very slow trajectories. The difference is that one bag file has trajectories with 0 velocities at all waypoints, while the other one has zero velocities at the beginning and end, and constant velocities in between. For both of these, we have computed by hand the expected accelerations and velocities, and they seem to be within reasonable limits. However, the arm oscillates both at the beginning and the end of the trajectories.
